### PR TITLE
Reimplement get_submodules_() without get_content_info()  (Reincarnated 6942)

### LINKED ISSUE #6940: `subdatasets` slow for no apparent reason
Scenario: dataset with two non-installed subdatasets and 40k top-level subdirectories:

```
% datalad -f disabled subdatasets 
datalad -f disabled subdatasets  91.01s user 11.94s system 100% cpu 1:42.54 total
% time git submodule
-dfa6d975ea8888ed33bf714c674371e1ab54e834 code/pipeline
-0c7f0b45140dde1d7291b15725e2016231a670b0 inputs/ukb
git submodule  0.02s user 0.02s system 105% cpu 0.035 total
% cat .gitmodules
[submodule "code/pipeline"]
        path = code/pipeline
        url = http://containers.ds.inm7.de/alias/cat
        datalad-id = 99300fb0-ea39-4fd2-b36d-2afc1e9ab7b2
[submodule "inputs/ukb"]
        path = inputs/ukb
        url = http://ukb.ds.inm7.de/alias/bids
        datalad-id = 14a5e5b6-d30d-11ea-8836-b4969157768c
% git submodule
-dfa6d975ea8888ed33bf714c674371e1ab54e834 code/pipeline
-0c7f0b45140dde1d7291b15725e2016231a670b0 inputs/ukb
```

So 35ms (git) vs 1m42s (datalad). :facepalm: 

And the reason seems to be in `GitRepo.get_submodules_()`:

https://github.com/datalad/datalad/blob/6b7a2f7cf480861ce44d162b4946f108acecf027/datalad/support/gitrepo.py#L2357-L2368

It calls `get_content_info()` to get *everything* about the dataset, only to discard anything that is not a `type='dataset'`. In this case of 2 subdatasets and 80k directories this is *highly* inefficient.

It seems to make sense to limit the call to `get_content_info()` to the paths obtain from `_parse_gitmodules()` already (or even a dedicated `git submodule` call).

`get_submodules_()` currently yields records like this:

```py
 {'gitmodule_datalad-id': '14a5e5b6-d30d-11ea-8836-b4969157768c',
  'gitmodule_name': 'inputs/ukb',
  'gitmodule_url': 'http://ukb.ds.inm7.de/alias/bids',
  'gitshasum': '0c7f0b45140dde1d7291b15725e2016231a670b0',
  'path': PosixPath('/home/mih/ukb-gource/ukb-derivatives/inputs/ukb'),
  'type': 'dataset'}
```

All properties prefixed with `gitmodule_` and the path are instantaneously provided by `_parse_gitmodules()`. The `type='dataset'` is pretty much implied and the `gitshasum` could come from `get_content_info()`, parameterized with the "intersection" of the reported submodule paths and the `paths` argument of `get_submodules_()`.

However, https://github.com/datalad/datalad/issues/6941 points out that `get_submodules_()` and `get_content_info()` are actually circular dependencies. So it would probably be better to implement the `gitshasum` query with a plain `ls-files --stage` call, instead of going through the full complexity of `get_content_info()`.

### PR DESCRIPTION
**Pull request by adswa**
A while ago, @mih proposed #6942. This effort was supposed to be super-seeded by #6974, which aimed to fix a short coming of #6942, namely an O(N²) performance hit in an edgecase ([when one provides paths to all subdatasets to a subdataset call](https://github.com/datalad/datalad/pull/6942/files#r947956976)). That fix is currently in draft mode, as it unexpectedly required more work than anticipated. @yarikoptic therefore stated in #6942 before it was closed

> @mih - I ran into need for more work on get_parent_paths since its goal is a bit different for matching paths within submodules not limiting some paths by other paths... needs a bit more work.
so I am ok with you solution, even if later to be improved later. But the tests I added in https://github.com/datalad/datalad/pull/6974 fails here and passes on master, so we would need to at least fix it up to not introduce regression. 

As I can still see the performance boost #6942 brought, I wanted to reincarnate the PR in its last state.
This is with ``master`` and on this branch with the same dataset @mih tested in #6942:

```
# on master
(gooey) adina@muninn in /tmp/ukb_cat on git:master
❱ time datalad subdatasets                                     
subdataset(ok): code/pipeline (dataset)
datalad subdatasets  177.08s user 16.30s system 99% cpu 3:13.39 total
# on this branch
(gooey) adina@muninn in /tmp/ukb_cat on git:master
❱ rehash
(gooey) adina@muninn in /tmp/ukb_cat on git:master
❱ datalad --version
datalad 0.17.9+158.g087ac2612
(gooey) adina@muninn in /tmp/ukb_cat on git:master
❱ time datalad subdatasets
subdataset(ok): code/pipeline (dataset)
datalad subdatasets  0.28s user 0.06s system 100% cpu 0.336 total
```

 I tested the tests that were failing before after merging ``master`` on my Debian system and on a Windows system, were they passed. Based on that, I'm slightly hopeful that our CI might turn green now as well :crossed_fingers: 
 
 I do see the performance hit in the edge case @yarikoptic discovered, though, still. Here are my timings in the human-connectome-project openaccess dataset:
 
 ```
 # on this branch
(gooey) adina@muninn in ~/repos/data/human-connectome-project-openaccess on git:master
❱ datalad --version; (builtin cd /home/adina/repos/data/human-connectome-project-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_(glob('HCP1200/*')))))" | wc -l; )

datalad 0.17.9+159.ge642f4159
1113
python -c   3.42s user 0.02s system 100% cpu 3.425 total
wc -l  0.00s user 0.00s system 0% cpu 3.425 total
# on master
(gooey) adina@muninn in ~/repos/data/human-connectome-project-openaccess on git:master
❱ datalad --version; (builtin cd /home/adina/repos/data/human-connectome-project-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_(glob('HCP1200/*')))))" | wc -l; )

datalad 0.17.9+153.g90cf7f374
1113
python -c   0.24s user 0.02s system 104% cpu 0.253 total
wc -l  0.00s user 0.00s system 0% cpu 0.252 total
 ```
 
 Nevertheless, I feel like the change so far *does* speed up the standard case considerably. If I don't explicitly provide paths to subdatasets, the timing difference is negligible:
 
 ```
# on master
(gooey) adina@muninn in ~/repos/data/human-connectome-project-openaccess on git:master
❱ datalad --version; (builtin cd /home/adina/repos/data/human-connectome-project-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_())))" | wc -l; ) 

datalad 0.17.9+153.g90cf7f374
1113
python -c   0.17s user 0.03s system 102% cpu 0.187 total
wc -l  0.00s user 0.00s system 0% cpu 0.187 total
# on this branch
(gooey) adina@muninn in ~/repos/data/human-connectome-project-openaccess on git:master
❱ rehash
(gooey) adina@muninn in ~/repos/data/human-connectome-project-openaccess on git:master
❱ datalad --version; (builtin cd /home/adina/repos/data/human-connectome-project-openaccess; time python -c "from datalad.support.gitrepo import *; from glob import glob; print('\n'.join(map(str,GitRepo('.').get_submodules_())))" | wc -l; ) 

datalad 0.17.9+159.ge642f4159
1113
python -c   0.18s user 0.03s system 107% cpu 0.197 total
wc -l  0.00s user 0.00s system 0% cpu 0.196 total
 ```
 
 
 Fixes #6940

### CODE REVIEW

**Comment by yarikoptic** [comment]
if you just reincarnated it, odd that those tests I have added do not fail -- but that is a good sign ;-)

**Comment by yarikoptic** [comment]
> Nevertheless, I feel like the change so far _does_ speed up the standard case considerably.

I don't mind this PR since detrimental effect for one usage pattern seems to be less that the performance boost for another, but I want to avoid claiming something to be standard whenever there is none "standard". Using globs in command line is IMHO no less common than pointing to a subpath. So this PR will make `datalad` faster for some people and slow for other.  And there is no "standard" way to use datalad (e.g. "not using globs is the *standard* way").

**Comment by adswa** [comment]
> I want to avoid claiming something to be standard whenever there is none "standard"

True, sorry - that was poorly phrased. 

I squashed the few fixes I did on top of the old branch into a single commit, and tuned the changelog.

**Review by bpoldrack — COMMENTED** [review]
Thanks, @adswa!

I have not yet looked into the next failure really. Have you?

**Comment by adswa** [comment]
> I have not yet looked into the next failure really. Have you?

I haven't as they looked completely unrelated to submodules: 

```
FAILED constraints/tests/test_cmdarg_validation.py::test_cmd_with_validation - AssertionError: assert [{'http://exa.../ '/dev/null'}] == [{'http://exa.../'/dev/null')}]
  At index 0 diff: {'http://example.com/': 'some/dir/file'} != {'http://example.com/': PosixPath('some/dir/file')}
  Full diff:
    [
  -  {'http://example.com/': PosixPath('some/dir/file')},
  ?                         ----------               -
  +  {'http://example.com/': 'some/dir/file'},
  -  {'file:///dev/null': PosixPath('/dev/null')},
  ?                       ----------           -
  +  {'file:///dev/null': '/dev/null'},
    ]
FAILED tests/test_download.py::test_download - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'download',
  'error_message': '404 Client Error: File not found for url: '
                   'http://127.0.0.1:35849/testfile.txt%20testfile2.txt',
  'exception': 404 Client Error: File not found for url: http://127.0.0.1:35849/testfile.txt%20testfile2.txt [download.py:__call__:315,http_url_operations.py:download:74,http_url_operations.py:download:69,models.py:raise_for_status:1021],
  'exception_traceback': '[download.py:__call__:315,http_url_operations.py:download:74,http_url_operations.py:download:69,models.py:raise_for_status:1021]',
  'message': 'download failure',
  'path': PosixPath('/tmp/datalad_temp_test_download81zvewl0/testfile.txt testfile2.txt'),
  'status': 'error',
  'status_code': 404,
  'url': 'http://127.0.0.1:35849/testfile.txt testfile2.txt'}]
```
But I can check if ``download file`` is somehow impacted.

File: changelog.d/pr-7189.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+### 🏎 Performance
+
+- Reimplement get_submodules_() without get_content_info() for substantial performance boosts especially for large datasets with few subdatasets. Originally proposed in 6942 by @mih, fixing #6940.  [PR #7189](https://github.com/datalad/datalad/pull/7189) (by [@adswa](https://github.com/adswa))
```
Comment (yarikoptic) [inline changelog.d/pr-7189.md:3]:
```suggestion
- Reimplement `get_submodules_()` without `get_content_info()` for substantial performance boosts especially for large datasets with few subdatasets. Note: might result for slightly slower performance when many subdataset paths are provided (e.g. via glob) to the function explicitly. Originally proposed in [PR #6942](https://github.com/datalad/datalad/pull/6942) by [@mih](https://github.com/mih), fixing [#6940](https://github.com/datalad/datalad/issues/6940).  [PR #7189](https://github.com/datalad/datalad/pull/7189) (by [@adswa](https://github.com/adswa))
```

**Comment by bpoldrack** [comment]
Ok, datalad-next failure unrelated, @yarikoptic "doesn't mind" and it doesn't prevent any further work on #6974 - let's merge then.

Thanks for digging this up, @adswa.

**Comment by yarikoptic-gitmate** [comment]
PR released in [`0.18.0`](https://github.com/datalad/datalad/releases/tag/0.18.0)

**Pull request merged**